qa/setup_packages.py

```diff
@@ -16,34 +16,433 @@
+            it is clamped to it
+        `cuda_version`: str, optional, default = None
+            Cuda version used for this query
+    """
+    def get_version(self, idx, cuda_version=None):
+        idx = self.clamp_index(idx, cuda_version)
+        return self.get_all_versions(cuda_version)[idx]
+
+    """Get all versions compatible with provided cuda_version
+
+        Parameters
+        ----------
+        `cuda_version`: str, optional, default = None
+            Cuda version used for this query
+    """
+    def get_all_versions(self, cuda_version=None):
+        raise NotImplementedError
+
+    """Obtains the number of available versions for given cuda_version
+
+        Parameters
+        ----------
+        `cuda_version`: str, optional, default = None
+            Cuda version used for this query
+    """
+    def get_num_of_version(self, cuda_version=None):
+        return len(self.get_all_versions(cuda_version))
+
+    """Obtains installation string that pip should accept for version at
+       a given index with a given cuda_version
+
+        Parameters
+        ----------
+        `idx`: int
+            Index of version to retrive. If index is beyond 0-num_ov_versions-1 range
+            it is clamped to it
+        `cuda_version`: str, optional, default = None
+            Cuda version used for this query
+    """
+    def get_install_string(self, idx, cuda_version=None):
+        raise NotImplementedError
+
+    """Obtains version string for version at a given index
+       with a given cuda_version
+
+        Parameters
+        ----------
+        `idx`: int
+            Index of version to retrive. If index is beyond 0-num_ov_versions-1 range
+            it is clamped to it
+        `cuda_version`: str, optional, default = None
+            Cuda version used for this query
+    """
+    def get_version_string(self, idx, cuda_version=None):
+        raise NotImplementedError
+
+    """Gets all installation string that pip should accept for a given
+       cuda version. Providing all of them to pip won't work, but each of
+       them should be a valid pip argument
+
+        Parameters
+        ----------
+        `cuda_version`: str, optional, default = None
+            Cuda version used for this query
+    """
+    def get_all_install_strings(self, cuda_version=None):
+        ret = []
+        for i in range(self.get_num_of_version(cuda_version)):
+            ret.append(self.get_install_string(i, cuda_version))
+        return " ".join(ret)
+
+    """Gets all version strings for with a given cuda_version
+
+        Parameters
+        ----------
+        `cuda_version`: str, optional, default = None
+            Cuda version used for this query
+    """
+    def get_all_configs(self, cuda_version=None):
+        ret = []
+        for i in range(self.get_num_of_version(cuda_version)):
+            ret.append(self.get_version(i, cuda_version))
+        return self.get_name(cuda_version) + ":\n\t" + "\n\t".join(ret)
```

Comment:
Why do you operate on strings here instead of lists? 
It feels to close to bash at this point.

Reply:
It is only to provide something pretty to print, like:
```
tensorflow-gpu:
     1.14
     1.15
...
```

Reply:
Anyway, reworked that as well.